sdk-exporter/trace/shared/src/main/scala/org/typelevel/otel4s/sdk/exporter/otlp/OtlpHttpSpanExporter.scala

```diff
@@ -0,0 +1,323 @@
+import fs2.Chunk
+import fs2.compression.Compression
+import fs2.io.net.Network
+import fs2.io.net.tls.TLSContext
+import org.http4s.EntityEncoder
+import org.http4s.Header
+import org.http4s.Headers
+import org.http4s.HttpVersion
+import org.http4s.Method
+import org.http4s.ProductId
+import org.http4s.Request
+import org.http4s.Response
+import org.http4s.Status
+import org.http4s.Uri
+import org.http4s.client.Client
+import org.http4s.client.middleware.{RetryPolicy => HttpRetryPolicy}
+import org.http4s.client.middleware.GZip
+import org.http4s.client.middleware.Retry
+import org.http4s.ember.client.EmberClientBuilder
+import org.http4s.headers.`User-Agent`
+import org.http4s.syntax.literals._
+import org.typelevel.ci._
+import org.typelevel.otel4s.sdk.BuildInfo
+import org.typelevel.otel4s.sdk.trace.data.SpanData
+import org.typelevel.otel4s.sdk.trace.exporter.SpanExporter
+
+import scala.concurrent.TimeoutException
+import scala.concurrent.duration._
+
+/** Exports spans via HTTP. Support `json` and `protobuf` encoding.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/protocol/exporter/]]
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/concepts/sdk-configuration/otlp-exporter-configuration/]]
+  */
+private final class OtlpHttpSpanExporter[F[_]: Temporal: Console] private (
+    client: Client[F],
+    config: OtlpHttpSpanExporter.Config
+) extends SpanExporter[F] {
+  import OtlpHttpSpanExporter.Encoding
+
+  val name: String = {
+    val headers = config.headers.mkString(
+      "headers={",
+      ",",
+      "}",
+      Headers.SensitiveHeaders
+    )
+
+    "OtlpHttpSpanExporter{" +
+      s"encoding=${config.encoding}, " +
+      s"endpoint=${config.endpoint}, " +
+      s"timeout=${config.timeout}, " +
+      s"gzipCompression=${config.gzipCompression}, " +
+      headers +
+      "}"
+  }
+
+  private implicit val spansEncoder: EntityEncoder[F, List[SpanData]] =
+    config.encoding match {
+      case Encoding.Json =>
+        import org.http4s.circe._
+        import io.circe.Printer
+        jsonEncoderWithPrinter(Printer.noSpaces)
+          .contramap[List[SpanData]](spans => ProtoEncoder.toJson(spans))
+
+      case Encoding.Protobuf =>
+        val content = Header.Raw(ci"Content-Type", "application/x-protobuf")
+        EntityEncoder.simple(content) { spans =>
+          Chunk.array(ProtoEncoder.toByteArray(spans))
+        }
+    }
+
+  def exportSpans[G[_]: Foldable](spans: G[SpanData]): F[Unit] = {
+    val request =
+      Request[F](Method.POST, config.endpoint, HttpVersion.`HTTP/1.1`)
+        .withEntity(spans.toList)
+        .putHeaders(config.headers)
+
+    client
+      .run(request)
+      .use(response => logBody(response).unlessA(response.status.isSuccess))
+      .timeoutTo(
+        config.timeout,
+        Temporal[F].raiseError(
+          new TimeoutException(
+            s"The export to [${config.endpoint}] has timed out after [${config.timeout}]"
+          )
+        )
```

Comment:
It needs to be something like this to make it lazy, otherwise the exception is computed eagerly which is very expensive.

```suggestion
        Temporal[F].unit >> Temporal[F].raiseError(
          new TimeoutException(
            s"The export to [${config.endpoint}] has timed out after [${config.timeout}]"
          )
        )
```

Reply:
Good point, thanks